cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -1005,6 +1007,325 @@ struct GroupedSumFactory {
+      double* other_means = reinterpret_cast<double*>(state.means_.mutable_data());
+      double* other_m2s = reinterpret_cast<double*>(state.m2s_.mutable_data());
+
+      arrow::internal::VisitSetBitRunsVoid(
+          array.buffers[0], array.offset + start_index,
+          std::min(max_length, batch.length - start_index),
+          [&](int64_t pos, int64_t len) {
+            for (int64_t i = 0; i < len; ++i) {
+              const int64_t index = start_index + pos + i;
+              const auto value = values[index];
+              sum[g[index]] += value;
+              square_sum[g[index]] += static_cast<uint64_t>(value) * value;
+              other_counts[g[index]]++;
+            }
+          });
+
+      for (int64_t i = 0; i < num_groups_; i++) {
+        if (other_counts[i] == 0) continue;
+
+        const double mean = static_cast<double>(sum[i]) / other_counts[i];
+        // calculate m2 = square_sum - sum * sum / count
+        // decompose `sum * sum / count` into integers and fractions
+        const int128_t sum_square = static_cast<int128_t>(sum[i]) * sum[i];
+        const int128_t integers = sum_square / other_counts[i];
+        const double fractions =
+            static_cast<double>(sum_square % other_counts[i]) / other_counts[i];
+        const double m2 = static_cast<double>(square_sum[i] - integers) - fractions;
+
+        other_means[i] = mean;
+        other_m2s[i] = m2;
+      }
+      RETURN_NOT_OK(this->Merge(std::move(state), group_id_mapping));
+    }
+    return Status::OK();
+  }
+
+  Status Merge(GroupedAggregator&& raw_other,
+               const ArrayData& group_id_mapping) override {
+    // Combine m2 from two chunks (see aggregate_var_std.cc)
+    auto other = checked_cast<GroupedVarStdImpl*>(&raw_other);
+
+    auto counts = reinterpret_cast<int64_t*>(counts_.mutable_data());
+    auto means = reinterpret_cast<double*>(means_.mutable_data());
+    auto m2s = reinterpret_cast<double*>(m2s_.mutable_data());
+
+    const auto* other_counts = reinterpret_cast<const int64_t*>(other->counts_.data());
+    const auto* other_means = reinterpret_cast<const double*>(other->means_.data());
+    const auto* other_m2s = reinterpret_cast<const double*>(other->m2s_.data());
+
+    auto g = group_id_mapping.GetValues<uint32_t>(1);
+    for (int64_t other_g = 0; other_g < group_id_mapping.length; ++other_g, ++g) {
+      if (other_counts[other_g] == 0) continue;
+      const double mean =
+          (means[*g] * counts[*g] + other_means[other_g] * other_counts[other_g]) /
+          (counts[*g] + other_counts[other_g]);
+      m2s[*g] += other_m2s[other_g] +
+                 counts[*g] * (means[*g] - mean) * (means[*g] - mean) +
+                 other_counts[other_g] * (other_means[other_g] - mean) *
+                     (other_means[other_g] - mean);
+      counts[*g] += other_counts[other_g];
+      means[*g] = mean;
+    }
```

Comment:
A pity this is duplicating the existing math from the scalar aggregate kernel.

How would you feel about factoring the underlying math in a simple `VarStdOp<ArrowType>` that you would feed values to? You would have one `VarStdOp` in the scalar aggregate kernel and `num_groups_` ones in the hash aggregate kernel.

That might be a bit different performance-wise because you would have an array-of-structures `std::vector<VarStdOp>` rather than a structure-of-arrays of the current three `BufferBuilder`, but I'm not sure it's really important here.

OTOH, `Consume` would not really benefit because the scalar aggregate kernel uses pairwise-summation for floating-point input.

cc @bkietz for advice

Reply:
Also @cyb70289

Reply:
Another option that Ben's mentioned would be to treat scalar aggregation as a hash aggregation with one group, though then we should immediately tackle the pairwise summation issue.

Reply:
https://issues.apache.org/jira/browse/ARROW-13451

Reply:
I'll see if I can get a comparison up before I'm out.

Reply:
It's unfortunately not very good. See #10813.

Reply:
I'll try refactoring it instead and compare the performance.